idaes/core/util/parameter_sweep.py

```diff
@@ -0,0 +1,728 @@
+        default=None,
+        domain=dict,
+        doc="Dict of options to pass to solver",
+    ),
+)
+
+
+@document_kwargs_from_configdict(CONFIG)
+class ParameterSweepBase:
+    """
+    IDAES base class for defining parameter sweep runners.
+
+    Thus base class defines a basic API for setting up parameter sweep studies
+    and allows the end-user to provide instructions for how to set up and run
+    the model to be studied using a set of callbacks.
+    """
+
+    def __init__(self, **kwargs):
+        self.config = CONFIG(kwargs)
+        self._results = OrderedDict()
+
+    @property
+    def results(self):
+        """
+        Returns OrderedDict containing the results from the parameter sweep.
+        """
+        return self._results
+
+    def execute_parameter_sweep(self):
+        """
+        Placeholder method for parameter sweep runners.
+
+        Developers should overload this with a method to execute the parameter sweep
+        using their preferred workflow manager.
+
+        Raises:
+            NotImplementedError
+        """
+        raise NotImplementedError(
+            "Derived classes should overload this method with a " "workflow manager."
+        )
+
+    def execute_single_sample(self, sample_id: int):
+        """
+        Executes a single run of the model using input values from sample_id.
+
+        Args:
+            sample_id: int indicating row in specification.samples to load into model
+                for run.
+
+        Returns:
+            results: results generated by collect_results callback
+            solved: bool indicating whether solver reported optimal convergence
+        """
+        model = self.get_initialized_model()
+
+        # Load sample values
+        self.set_input_values(model, sample_id)
+
+        # Solve model and capture output
+        solver = self._get_solver()
+
+        # Try/except to catch any critical failures that occur
+        try:
+            status, run_stats = self.run_model(model, solver)
+
+            solved = check_optimal_termination(status)
+            if not solved:
+                _log.error(f"Sample: {sample_id} failed to converge.")
+
+            # Compile Results
+            results = self.collect_results(model, status, run_stats)
+        except:  # pylint: disable=bare-except
+            # Catch any Exception for recourse
+            results, solved = self.execute_recourse(model)
+
+        return results, solved
+
+    def get_initialized_model(self):
+        """
+        Get instance of model to be run by calling build_model callback.
+
+        Returns:
+            Pyomo model to be run
+        """
+        if self.config.build_model is None:
+            raise ConfigurationError(
+                "Please specify a method to construct the model of interest."
+            )
+        args = self.config.build_model_arguments
+        if args is None:
+            args = {}
+
+        model = self.config.build_model(**args)
```

Comment:
Should there be an option to send sampled parameter values to the `build_model` function? I could imagine this being useful if we want to set parameter values before initializing the model. (Or maybe this should be done by implementing a custom `run_model` function that calls `model.initialize()` before solving?)

Reply:
That was part of my intention for the `run_model` method; `build_model` would do everything up to setting values, and then `run_model` would do anything that needed to follow.

Reply:
I think that makes sense